engine/access/rpc/backend/backend_events.go

```diff
@@ -106,10 +118,95 @@ func (b *backendEvents) GetEventsForBlockIDs(
 		blockHeaders = append(blockHeaders, header)
 	}
 
-	// forward the request to the execution node
-	return b.getBlockEventsFromExecutionNode(ctx, blockHeaders, eventType, requiredEventEncodingVersion)
+	return b.getBlockEvents(ctx, blockHeaders, eventType, requiredEventEncodingVersion)
 }
 
+// getBlockEvents retrieves events for all the specified blocks that have the given type
+// It gets all events available on storage, and requests the rest to an execution node.
+func (b *backendEvents) getBlockEvents(
+	ctx context.Context,
+	blockHeaders []*flow.Header,
+	eventType string,
+	requiredEventEncodingVersion entities.EventEncodingVersion,
+) ([]flow.BlockEvents, error) {
+	localResponse, missingHeaders, err := b.getBlockEventsFromStorage(ctx, blockHeaders, eventType, requiredEventEncodingVersion)
+	if err != nil {
+		return nil, err
+	}
+
+	enResponse, err := b.getBlockEventsFromExecutionNode(ctx, missingHeaders, eventType, requiredEventEncodingVersion)
+	if err != nil {
+		return nil, err
+	}
+
+	// sort ascending by block height
+	// Note: this may not match the order of the original request for clients using GetEventsForBlockIDs
+	// that provide out of order block IDs
+	response := append(localResponse, enResponse...)
+	sort.Slice(response, func(i, j int) bool {
+		return response[i].BlockHeight < response[j].BlockHeight
+	})
+
+	return response, nil
+}
+
+// getBlockEventsFromStorage retrieves events for all the specified blocks that have the given type
+// from the local storage
+func (b *backendEvents) getBlockEventsFromStorage(
+	ctx context.Context,
+	blockHeaders []*flow.Header,
+	eventType string,
+	requiredEventEncodingVersion entities.EventEncodingVersion,
+) ([]flow.BlockEvents, []*flow.Header, error) {
+	target := flow.EventType(eventType)
+
+	missing := make([]*flow.Header, 0)
+	resp := make([]flow.BlockEvents, 0)
+	for _, header := range blockHeaders {
+		if ctx.Err() != nil {
+			return nil, nil, rpc.ConvertError(ctx.Err(), "failed to get events", codes.Canceled)
+		}
+
+		events, err := b.events.ByBlockID(header.ID())
+		if err != nil {
+			if errors.Is(err, storage.ErrNotFound) {
```

Comment:
I'm 90% sure `ErrNotFound` is only returned if events for the block have not been indexed. double check and add a comment explaining